zbus_names/src/bus_name.rs

```diff
@@ -213,49 +212,39 @@ impl<'s> TryFrom<&'s str> for BusName<'s> {
     }
 }
 
-impl TryFrom<String> for BusName<'_> {
+impl<'s> TryFrom<&'s str> for BusName<'s> {
+    type Error = Error;
+
+    fn try_from(value: &'s str) -> Result<Self> {
+        let value: Str<'s> = value.into();
+        value.try_into()
+    }
+}
+
+impl<'s> TryFrom<String> for BusName<'s> {
     type Error = Error;
 
     fn try_from(value: String) -> Result<Self> {
-        Ok(match BusName::try_from(value.as_str())? {
-            BusName::Unique(_) => BusName::Unique(UniqueName::from_string_unchecked(value)),
-            BusName::WellKnown(_) => {
-                BusName::WellKnown(WellKnownName::from_string_unchecked(value))
-            }
-        })
+        let value: Str<'s> = value.into();
+        value.try_into()
     }
 }
 
-impl TryFrom<Arc<str>> for BusName<'_> {
+impl<'s> TryFrom<Arc<str>> for BusName<'s> {
     type Error = Error;
 
     fn try_from(value: Arc<str>) -> Result<Self> {
-        // FIXME: Same as `str` impl above. Use a macro to avoid code duplication.
-        match UniqueName::try_from(value.clone()) {
-            Err(Error::InvalidUniqueName(unique_err)) => match WellKnownName::try_from(value) {
-                Err(Error::InvalidWellKnownName(well_known_err)) => {
-                    Err(Error::InvalidBusName(unique_err, well_known_err))
-                }
-                Err(e) => Err(e),
-                Ok(name) => Ok(BusName::WellKnown(name)),
-            },
-            Err(e) => Err(e),
-            Ok(name) => Ok(BusName::Unique(name)),
-        }
+        let value: Str<'s> = value.into();
+        value.try_into()
```

Comment:
The BusName changes are not related to the de-duplicating ones, hence separate commits please. The "and" in the commit shortlog is typically a telltale sign of multiple logical changes.